### PR TITLE
feat-#178: added admin UI to frontend

### LINKED ISSUE #178: [FEATURE] Adding admin functionality to the frontend
### Description

1 Admin routes shouldn't be accessible when the URL is directly entered.
2 owner of the post should be able to edit and delete his own post
3 admin should be able to see the edit and delete post option like the 2nd point like post owner. At the same time has his dedicated window using 3rd point.

### Use Case

This feature will complete the admin functionality to the frontend.

### Proposed Solution

_No response_

### Additional Information

_No response_

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by devsharmagit**
## Summary

Added admin UI to frontend

## Description

This PR adds all the admin functionality and UI to the frontend.
Checkout by logging in as admin.
/admin/user and /admin/blog route is implemented correctly.
A button in the header layout is also added to navigate to the admin dash board.

https://github.com/krishnaacharyaa/wanderlust/assets/109821078/7c5542ff-5108-4c60-8647-8710dea4635e

## Issue(s) Addressed

Closes #178 

## Prerequisites

- [x] Have you followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions)?

### CODE REVIEW

**Comment by krishnaacharyaa** [comment]
@devsharmagit can you kindly rebase to upstream's main

**Comment by krishnaacharyaa** [comment]
@devsharmagit can you kindly rebase to upstream's main

**Comment by krishnaacharyaa** [comment]
@devsharmagit kindly finish this

**Comment by devsharmagit** [comment]
@krishnaacharyaa 
I am back to normal. I will commit the changes today night.

File: frontend/src/pages/admin-users.tsx [hunk 1]
```diff
@@ -1,22 +1,88 @@
+import axiosInstance from '@/helpers/axios-instance';
+import { useEffect, useState } from 'react';
+import { toast } from 'react-toastify';
+import 'react-toastify/dist/ReactToastify.css';
+
+interface User {
+  _id: string;
+  fullName: string;
+  role: 'ADMIN' | 'USER';
+  email: string;
```
Comment (krishnaacharyaa) [inline frontend/src/pages/admin-users.tsx:10]: This guy shouldn't he use the enum defined below? 

Can you kindly replace all the interface to types

File: frontend/src/pages/admin-users.tsx [hunk 2]
```diff
@@ -1,22 +1,88 @@
...
+                <div className="flex flex-col gap-[10px] ">
+                  <p className="text-base font-medium text-light-title dark:text-dark-title">
+                    {user?.fullName}
+                  </p>
+                  <p className="text-base font-medium text-light-description dark:text-dark-description">
+                    {user?.email}
+                  </p>
+                </div>
+                {user.role === 'ADMIN' && (
+                  <button
```
Comment (krishnaacharyaa) [inline frontend/src/pages/admin-users.tsx:68]: Use enum instead, kindly replace all the places this is used to use enum

**Review by krishnaacharyaa — CHANGES_REQUESTED** [review]
Please just have a video in PR description showcasing the functionality

**Review by krishnaacharyaa — CHANGES_REQUESTED** [review]
1. The enum role  as far as I know, he is been used everywhere, Can we lift it up 
And the User type isn't it already defined? I feel it should go to its proper types folder, 
Because it is not specific to only `admin-users` right? 

2. And we need to call the clean redis function as well, when ever any post is deleted or edited.

3. I guess editing is not functional, if it is not create an issue so that someone else can work on this, so that we don't miss.

I can get this in, if you can raise another PR for the refactoring (1)
We need (2) changes and (3) to get this in, meaning merge, (1) we can take it seperately

cc: @devsharmagit 
Thank you for all the support :)

**Comment by devsharmagit** [comment]
@krishnaacharyaa 
I have never used Redis. I will first go through some resources on YouTube to understand.
I will commit the required changes tomorrow (Saturday).

I will raise an issue related to editing the place and eventually work on it and complete it.
I will also raise the issue of refactoring the code as mentioned in 1.

**Comment by krishnaacharyaa** [comment]
> I have never used Redis. I will first go through some resources on YouTube to understand.

Just use the existing util function @devsharmagit to clean up the redis

**Comment by devsharmagit** [comment]
@krishnaacharyaa 
i will use it and push the changes in few hours.

apart from that everything is fine from your side?

**Comment by devsharmagit** [comment]
image
is this the function that i need to call?

and what is the argument that i need to pass?

@krishnaacharyaa

**Comment by krishnaacharyaa** [comment]
> image is this the function that i need to call?
> 
> and what is the argument that i need to pass?
> 
> @krishnaacharyaa

Yes. Figure out from other places it is being used
@devsharmagit

**Comment by devsharmagit** [comment]
okay @krishnaacharyaa !

**Comment by devsharmagit** [comment]
@krishnaacharyaa 
I have pushed the code.

**Comment by krishnaacharyaa** [comment]
@devsharmagit kindly raise 1 and 3 and comment here, so that i close this

File: backend/controllers/posts-controller.js [hunk 3]
```diff
@@ -135,8 +135,11 @@ export const updatePostHandler = async (req, res) => {
     if (!updatedPost) {
       return res.status(HTTP_STATUS.NOT_FOUND).json({ message: RESPONSE_MESSAGES.POSTS.NOT_FOUND });
     }
-
-    res.status(HTTP_STATUS.OK).json(updatedPost);
+    // invalidate the redis cache
+    await deleteDataFromCache(REDIS_KEYS.ALL_POSTS),
+      await deleteDataFromCache(REDIS_KEYS.FEATURED_POSTS),
```
Comment (krishnaacharyaa) [inline backend/controllers/posts-controller.js:140]: It should ideally format, I'm confused, why isn't it working for you !!
we have a husky hook which actually formats the code, 
Isn't it working for you?

File: backend/controllers/posts-controller.js [hunk 4]
```diff
@@ -152,7 +155,11 @@ export const deletePostByIdHandler = async (req, res) => {
     }
     await User.findByIdAndUpdate(post.authorId, { $pull: { posts: req.params.id } });
 
-    res.status(HTTP_STATUS.OK).json({ message: RESPONSE_MESSAGES.POSTS.DELETED });
+    // invalidate the redis cache
+    await deleteDataFromCache(REDIS_KEYS.ALL_POSTS),
```
Comment (krishnaacharyaa) [inline backend/controllers/posts-controller.js:159]: Same here

File: backend/controllers/posts-controller.js (hunk 3 above)
Comment (devsharmagit) [reply backend/controllers/posts-controller.js:140]: i guess its not working for me.
i will manually do "npm run format" in both frontend and backend and push the code

Comment (krishnaacharyaa) [reply backend/controllers/posts-controller.js:140]: Sure @devsharmagit

**Comment by devsharmagit** [comment]
@krishnaacharyaa 

global user type in frontend folder -  #418 https://github.com/krishnaacharyaa/wanderlust/issues/418
editing the posts - #419 https://github.com/krishnaacharyaa/wanderlust/issues/419

**Review by krishnaacharyaa — APPROVED** [review]
LGTM, thank you @devsharmag99 for the efforts

**Comment by devsharmagit** [comment]
@krishnaacharyaa 
actually @devsharmag99  is my company's github account. I forgot to switch before contributing

**Comment by krishnaacharyaa** [comment]
Oooh okay cool hai no problem @devsharmagit

**Pull request merged**